src/obj_char_transfer.c

```diff
@@ -0,0 +1,872 @@
+        }
+        sObjCharTransferTasksManager->blockSizeMain = ObjCharTransfer_GetBlockSizeFromMode(modeMain);
+        sObjCharTransferTasksManager->blockSizeSub = ObjCharTransfer_GetBlockSizeFromMode(modeSub);
+        GX_SetOBJVRamModeChar(modeMain);
+        GXS_SetOBJVRamModeChar(modeSub);
+        int numBlocksMain = ObjCharTransfer_CalcBlockNumLimit(template->sizeMain, sObjCharTransferTasksManager->blockSizeMain);
+        int numBlocksSub = ObjCharTransfer_CalcBlockNumLimit(template->sizeSub, sObjCharTransferTasksManager->blockSizeSub);
+        ObjCharTransfer_InitBlocksTransferBuffers(numBlocksMain, numBlocksSub, template->heapId);
+    }
+}
+
+void ObjCharTransfer_Destroy(void) {
+    if (sObjCharTransferTasksManager != NULL) {
+        ObjCharTransfer_FreeBlockTransferBuffer(sObjCharTransferTasksManager->blockBufMain);
+        ObjCharTransfer_FreeBlockTransferBuffer(sObjCharTransferTasksManager->blockBufSub);
+        resetAllTransferTasks();
+        FreeToHeap(sObjCharTransferTasksManager->tasks);
+        FreeToHeap(sObjCharTransferTasksManager);
+        sObjCharTransferTasksManager = NULL;
+    }
+}
+
+void ObjCharTransfer_ClearBuffers(void) {
+    sObjCharTransferTasksManager->offsetMain = 0;
+    sObjCharTransferTasksManager->offsetSub = 0;
+    ObjCharTransfer_ClearBlockBuf(sObjCharTransferTasksManager->blockBufMain);
+    ObjCharTransfer_ClearBlockBuf(sObjCharTransferTasksManager->blockBufSub);
+    ObjCharTransfer_GetVramCapacityByBank();
+}
+
+void ObjCharTransfer_SetReservedRegion(u32 offset, u32 size, NNS_G2D_VRAM_TYPE vram) {
+    int newOffset;
+    int newSize;
+    if (vram == NNS_G2D_VRAM_TYPE_2DMAIN) {
+        boundsFixOffsetAndSize(sObjCharTransferTasksManager->freeSizeMain, offset, size, &newOffset, &newSize);
+        if (newSize > 0) {
+            reserveTransferBlocksByVramOffsetAndSize(NNS_G2D_VRAM_TYPE_2DMAIN, newOffset, 0, newSize, 0);
+        }
+    } else {
+        boundsFixOffsetAndSize(sObjCharTransferTasksManager->freeSizeSub, offset, size, &newOffset, &newSize);
+        if (newSize > 0) {
+            reserveTransferBlocksByVramOffsetAndSize(NNS_G2D_VRAM_TYPE_2DSUB, 0, newOffset, 0, newSize);
+        }
+    }
+}
+
+BOOL ObjCharTransfer_CreateTaskAndDoTransferFromTemplate(const ObjCharTransferTaskTemplate *template) {
+    GF_ASSERT(ObjCharTransfer_TaskExistsByID(template->id) != TRUE);
+    ObjCharTransferTask *sub = ObjCharTransfer_GetFreeTask();
+    if (sub == NULL) {
+        GF_ASSERT(FALSE);
+        return FALSE;
+    }
+    if (!ObjCharTransferTask_InitFromTemplate(template, sub)) {
+        return FALSE;
+    }
+    if (!ObjCharTransferTask_ReserveAndTransfer(sub)) {
+        ObjCharTransfer_ResetTransferTasksByResID(sub->resId);
+        return FALSE;
+    }
+    ++sObjCharTransferTasksManager->num;
+    return TRUE;
+}
+
+BOOL ObjCharTransfer_CreateTaskAndDoTransferFromTemplate_UpdateMappingTypeFromHW(const ObjCharTransferTaskTemplate *template) {
```

Comment:
what is a HW, hardware?

Reply:
yes, "HW" represents "Hardware" elsewhere in the project, including the SDK